crates/sui-bridge/src/types.rs

```diff
@@ -1,27 +1,162 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use std::collections::HashMap;
-
-use fastcrypto::secp256k1::{Secp256k1PublicKey, Secp256k1Signature};
+use crate::crypto::{BridgeAuthorityPublicKey, BridgeAuthoritySignInfo, BridgeAuthoritySignature};
+use crate::{abi::EthBridgeEvent, events::SuiBridgeEvent};
+use ethers::types::Address as EthAddress;
+use fastcrypto::hash::HashFunction;
+use fastcrypto::hash::Keccak256;
+use serde::{Deserialize, Serialize};
+use shared_crypto::intent::IntentScope;
+use std::collections::{BTreeMap, HashMap};
+use sui_types::digests::Digest;
+use sui_types::error::SuiResult;
+use sui_types::message_envelope::{Envelope, Message, VerifiedEnvelope};
 use sui_types::multiaddr::Multiaddr;
+use sui_types::{base_types::SUI_ADDRESS_LENGTH, committee::EpochId};
 
-pub type BridgeCommitteePublicKey = Secp256k1PublicKey;
-pub type BridgeCommitteeSignature = Secp256k1Signature;
-
-#[derive(Debug, Eq, PartialEq)]
-pub struct CommitteeMember {
-    pub pubkey: BridgeCommitteePublicKey,
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct BridgeAuthority {
+    pub pubkey: BridgeAuthorityPublicKey,
     pub voting_power: u64,
     pub bridge_network_address: Multiaddr,
     pub is_blocklisted: bool,
 }
 
-#[derive(Debug, Eq, PartialEq)]
+// A static Bridge committee implementation
+#[derive(Debug)]
 pub struct BridgeCommittee {
-    pub members: Vec<CommitteeMember>,
+    pub members: BTreeMap<BridgeAuthorityPublicKey, BridgeAuthority>,
+}
+
+impl BridgeCommittee {
+    pub fn is_active_member(&self, member: &BridgeAuthorityPublicKey) -> bool {
+        self.members.contains_key(member) && !self.members.get(member).unwrap().is_blocklisted
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum BridgeEventType {
+    TokenTransfer = 0,
+    UpdateCommitteeBlocklist = 1,
+    EmergencyButton = 2,
+}
+
+pub const SUI_TX_DIGEST_LENGTH: usize = 32;
+pub const ETH_TX_HASH_LENGTH: usize = 32;
+
+pub const BRIDGE_MESSAGE_PREFIX: &[u8] = b"SUI_NATIVE_BRIDGE";
+
+#[derive(Copy, Clone)]
+pub enum BridgeChainId {
+    Sui = 0,
+    Eth = 1,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum TokenId {
+    Sui = 0,
+    BTC = 1,
+    ETH = 2,
+    USDC = 3,
+    USDT = 4,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum BridgeEvent {
+    Sui(SuiBridgeEvent),
+    Eth(EthBridgeEvent),
+}
+
+impl BridgeEvent {
+    /// Convert to message bytes that are verified in Move and Solidity
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        // Add prefix
+        bytes.extend_from_slice(BRIDGE_MESSAGE_PREFIX);
+        match self {
+            BridgeEvent::Sui(SuiBridgeEvent::SuiToEthTokenBridgeV1(e)) => {
+                // Add message type
```

Comment:
we can't use bcs because solidity does not have it. This is an encoding scheme both move and solidity can construct

Reply:
note that everything here is fixed size, so the bcs serialization should result in the same thing (assuming enums use #[repr(u8)])

Reply:
It's not exactly the same - we still need the address length. 

I don't see a fundamentally difference in terms of different encoding schemes, the goal is to have one that is not error prone and easy to implement in both move, solidity and rust. @patrickkuo  @georgedigkas  could you weight in here (I think it may add a bit complexity in solidity)

Reply:
Just to clarify - I didn't mean to use bcs in solidity, but that the same byte array we generate in the current code  can be generated, in rust, using bcs::to_bytes() (again, assuming we don't have variable size parameters)

Reply:
why does it have to be implemented as enum? 

bcs is probably easier to reason about, look at this example, bcs encoding should work: 

```
let e = SuiToEthBridgeEventV1 {
                nonce: 1,
                sui_address: SuiAddress::random_for_testing_only(),
                eth_address: EthAddress::random(),
                sui_tx_digest: TransactionDigest::random(),
                sui_tx_event_index: 1,
                token_id: TokenId::Sui,
                amount: 100,
            };
return bcs::to_bytes(e).map_err(...)
```